python/pywatchman/aioclient.py

```diff
@@ -0,0 +1,286 @@
+# Copyright 2014-present Facebook, Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice,
+#    this list of conditions and the following disclaimer.
+#
+#  * Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+#  * Neither the name Facebook nor the names of its contributors may be used to
+#    endorse or promote products derived from this software without specific
+#    prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS'
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import asyncio
+import os
+import subprocess
+import threading
+
+try:
+    from . import bser
+except ImportError:
+    from . import pybser as bser
+
+from . import compat, encoding
+
+from . import CommandError, SocketTimeout, WatchmanError
+
+
+# 2 bytes marker, 1 byte int size, 8 bytes int64 value
+SNIFF_LEN = 13
+
+
+class AsyncTransport(object):
+    '''Communication transport to the Watchman Service.'''
+
+    def close(self):
+        raise NotImplementedError()
+
+    async def read(self, size):
+        raise NotImplementedError()
+
+    async def write(self, buf):
+        raise NotImplementedError()
+
+
+class AsyncUnixSocketTransport(AsyncTransport):
+    '''Local Unix domain socket transport supporting asyncio.'''
+
+    def __init__(self, sockname, loop):
+        self.sockname = sockname
+        self.reader = None
+        self.writer = None
+        init_task = asyncio.ensure_future(self.__init_socket(), loop=loop)
+        loop.run_until_complete(init_task)
```

Comment:
I actually created a class method constructor that does the initialization instead. The constructor is now more decoupled from specifics of the Transports and Codecs.